packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { maxSatisfying } from 'semver';
+
+import { ArrayExt } from '@lumino/algorithm';
+
+import { PromiseDelegate } from '@lumino/coreutils';
+
+import { Debouncer } from '@lumino/polling';
+
+const TIMEOUT = 300;
+
+interface IRequest<T> {
+  name: string;
+  semVer: string;
+  result: PromiseDelegate<T | undefined>;
+  timestamp: number | null;
+}
+
+/**
+ * A registry using semver ranges to retrieve values.
+ *
+ * Each object in the registry is associated with a name and a semantic
+ * version. Objects can be retrieved by giving the name and a semantic version
+ * range, and the maximum satisfying version of the object will be returned.
+ *
+ * Object retrieval may be delayed for a short timeout if no matching object
+ * can be found. This is to allow a short window for objects to be registered
+ * before get requests are resolved as unsatisfied. Practically, this means
+ * that initial object registration can happen concurrent with the first
+ * requests.
+ *
+ */
+export class SemVerRegistry<T> {
+  /**
+   * Register a object with a given name and version.
+   */
+  register(options: SemVerRegistry.IRegistryData<T>): void {
+    const { name, version, object } = options;
+    let versions = this._registry.get(name);
+    if (versions === undefined) {
+      versions = new Map<string, T>();
+      this._registry.set(name, versions);
+    }
+
+    if (versions.has(version)) {
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._debounce.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(
+    name: string,
+    semVer: string,
+    isKernelRestore: boolean
```

Comment:
This shouldn't know anything about the concept of kernels. How about this parameter be a timeout (in ms) on the request, defaulting to 0? Then the thing calling get() can say to use a specific timeout on the request, like 500ms. Then this would mirror standard socket apis which have a timeout parameter.